h/views/api/helpers/url_template.py

```diff
@@ -0,0 +1,43 @@
+import re
+
+from pyramid.interfaces import IRoutesMapper
+
+
+def route_url_template(request, route_name: str) -> str:
+    """
+    Generate a URL template for a Pyramid route.
+
+    URL templates have the format "https://example.com/users/:username" where
+    ":username" is a dynamic parameter.
+    """
+
+    # Route lookup code taken from `request.route_url` implementation.
+    routes = request.registry.getUtility(IRoutesMapper)
+    route = routes.get_route(route_name)
+
+    if route is None:
+        msg = f"Route '{route_name}' not found"
+        raise ValueError(msg)
+
+    # Route patterns are strings like "/users/{username}/items/{item}". Extract
+    # the names of the dynamic parts.
+    #
+    # The dynamic parts can contain character set restrictions and also lengths,
+    # resulting in patterns like "/api/annotations/{id:[A-Za-z0-9_-]{20,22}}"
+    # which can't be parsed with a regex. These more complex cases are handled
+    # below by catching the `KeyError` and retrying.
+    param_names = re.findall(r"\{([A-Za-z0-9_.]+)\}", route.pattern)
+
+    def placeholder(name):
+        # nb. This assumes that the parameter name contains only URL-safe
+        # characters which will not be affected by URL encoding.
+        return f":{name}"
+
+    params = {name: placeholder(name) for name in param_names}
+
+    while True:
+        try:
+            return request.route_url(route_name, **params)
+        except KeyError as exc:
+            param = exc.args[0]
+            params[param] = placeholder(param)
```

Comment:
This doesn't protect against an infinite loop if `KeyError` was raised for a reason other than a missing param. I don't know of a different way to trigger a `KeyError`. The assumption is that if this does happen, we'll notice soon enough and fix it.